src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,53 +12,57 @@ function calc_error_norms(func, u, t, analyzer,
+  FT = eltype(l2_errors[1])
+  l2_errors = reshape(reinterpret(FT, l2_errors), nvars, :) |> Matrix
+  linf_errors = reshape(reinterpret(FT, linf_errors), nvars, :) |> Matrix
+
+  # Collect local error norms of all elements on root process
   if mpi_isroot()
-    l2_error   = convert(typeof(l2_error),   global_l2_error)
-    linf_error = convert(typeof(linf_error), global_linf_error)
+    global_l2_errors = zeros(eltype(l2_errors), nvars, cache.mpi_cache.n_elements_global)
+    global_linf_errors = similar(global_l2_errors)
+
+    n_elements_by_rank = parent(cache.mpi_cache.n_elements_by_rank) # convert OffsetArray to Array
+    l2_buf = MPI.VBuffer(global_l2_errors, nvars*n_elements_by_rank)
+    linf_buf = MPI.VBuffer(global_linf_errors, nvars*n_elements_by_rank)
+    MPI.Gatherv!(l2_errors, l2_buf, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, linf_buf, mpi_root(), mpi_comm())
   else
-    l2_error   = convert(typeof(l2_error),   NaN * global_l2_error)
-    linf_error = convert(typeof(linf_error), NaN * global_linf_error)
+    MPI.Gatherv!(l2_errors, nothing, mpi_root(), mpi_comm())
+    MPI.Gatherv!(linf_errors, nothing, mpi_root(), mpi_comm())
   end
 
-  l2_error = @. sqrt(l2_error / total_volume)
+  # Aggregate element error norms on root process
+  if mpi_isroot()
+    # Convert from Matrix to Vector of SVectors
+    global_l2_errors = reinterpret(T, vec(global_l2_errors))
+    global_linf_errors = reinterpret(T, vec(global_linf_errors))
 
-  return l2_error, linf_error
-end
+    # Aggregate element errors
 
+    # global_l2_errors is a Base.ReinterpretArray, by accessing it via [:] we convert it into a
+    # proper Vector so that sum always produces the same result as in the serial case
+    l2_error = sum(global_l2_errors[:])
+    linf_error = reduce((x, y) -> max.(x, y), global_linf_errors)
 
-function integrate_via_indices(func::Func, u,
-                               mesh::ParallelTreeMesh{2}, equations, dg::DGSEM, cache,
-                               args...; normalize=true) where {Func}
-  # call the method accepting a general `mesh::TreeMesh{2}`
-  # TODO: MPI, we should improve this; maybe we should dispatch on `u`
-  #       and create some MPI array type, overloading broadcasting and mapreduce etc.
-  #       Then, this specific array type should also work well with DiffEq etc.
-  local_integral = invoke(integrate_via_indices,
-    Tuple{typeof(func), typeof(u), TreeMesh{2}, typeof(equations),
-          typeof(dg), typeof(cache), map(typeof, args)...},
-    func, u, mesh, equations, dg, cache, args..., normalize=normalize)
-
-  # OBS! Global results are only calculated on MPI root, all other domains receive `nothing`
-  global_integral = MPI.Reduce!(Ref(local_integral), +, mpi_root(), mpi_comm())
-  if mpi_isroot()
-    integral = convert(typeof(local_integral), global_integral[])
+    # For L2 error, divide by total volume
+    total_volume = mesh.tree.length_level_0^ndims(mesh)
```

Comment:
Why can't you use the `total_volume` function here?

Reply:
That's from the old code. But it works with the `total_volume` function too, so I changed it.